### PR TITLE
[FEATURE] add library init routine

### LINKED ISSUE #352: SEGFAULT in xerces libraries when removing default instance of SvmTheoreticalSpectrumGenerator [352]
Submitted by aiche on 2011-09-21 13:48:56

If you remove the default instance of SvmTheoreticalSpectrumGenerator in SvmTheoreticalSpectrumGenerator.C some tests fail with segfaults caused by xerces methods like 

xercesc_3_0::XMLString::operator new
xercesc_3_0::XMLString::transcode

until we know how to fix this we will leave the default instance in the class.

### PR DESCRIPTION
**Pull request by hroest**
fixes #352

### CODE REVIEW

File: src/openms/include/OpenMS/CONCEPT/Init.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <OpenMS/CONCEPT/Exception.h>
+
+/**
+    @brief Initialization procedures
+
+    Part of the library needs to be initialized to be used properly, these
+    procedures need to be run at startup (and preferably only once).
+
+*/
+
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CONCEPT/Init.h]: don't we need an: extern xerces_init xinit; ?

File: src/openms/source/CONCEPT/Init.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        xercesc::XMLPlatformUtils::Initialize();
+      }
+
+      ~xerces_init() 
+      {
+        xercesc::XMLPlatformUtils::Terminate();
+      }
+
+    };
+    xerces_init xinit;
```
Comment (hroest) [inline src/openms/source/CONCEPT/Init.cpp]: maybe add const?

File: src/openms/include/OpenMS/CONCEPT/Init.h (hunk 1 above)
Comment (hroest) [reply src/openms/include/OpenMS/CONCEPT/Init.h]: only if we want to expose it to external callers, right? which we dont want to?

Global variables have global scope, and can be used anywhere in the program. Like functions, you must use a forward declaration (via keyword extern) to use a global variable defined in another file. By default, non-const global variables have external linkage.Jun 19, 2007

so maybe we should make it const

Comment (timosachsenberg) [reply src/openms/include/OpenMS/CONCEPT/Init.h]: "By default, non-const variables declared outside of a block are assumed to be external. However, const variables declared outside of a block are assumed to be internal."
Yes I guess we need const or static to declare internal linkage.

Comment (hroest) [reply src/openms/include/OpenMS/CONCEPT/Init.h]: I have just added const on github, lets see whether the tests work

**Comment by hroest** [comment]
what happened to the windows build?

**Comment by jpfeuffer** [comment]
Hmm unfortunately it did not upload to CDash. I increased the retention time of the build logs for CI on Jenkins (which was the reason for the deletion).

**Comment by jpfeuffer** [comment]
rebuild win

**Comment by jpfeuffer** [comment]
probably does not work after it was merged

**Comment by timosachsenberg** [comment]
hmm was the header file added to the cmake?

**Pull request merged**